WIP: Support parsing clispecs as yaml

scottwittenburg opened the pull request:
Currently commands that operate on specs, like `spack install` for example, might or might not additionally allow users to provide the `-f <path-to-spec.yaml>` argument as an alternate way to identify specs.  Using the `-f` option is nice because it removes all ambiguity as to what will be installed, or the spec for which a buildcache will be created, etc.

Instead of incrementally updating all the various commands where this `-f <path-to-spec.yaml>` functionality might be useful (see, for example, #10912), this PR represents an attempt to support it in one go for all commands simultaneously.

At a high level, this PR attempts to let the `spack.cmd.parse_specs(...)` method do what it has always done, i.e. assume the positional arguments are clispecs.  If that fails, however, then it additionally tries to see if the positional arguments are readable files representing concrete specs.  So for example, while you can still say:

```
spack install pkgconf bzip2
```

You could also now concretize the `pkgconf` and `bzip2` specs and write those out to the filesystem before saying:

```
spack install /tmp/pkgconf.yaml /tmp/bzip2.yaml
```

In order to unify the way commands handle specs on the command line, the first commit in this PR creates a new common command line argument, `specs`, as the `REMAINDER` of the positional args, and any commands that called `parse_specs()`, now use that common arg.  This also allows documenting the limitations discussed below in a single location, instead of in every command that operates on specs.

There are still some unanswered questions here though.

1. Previously the `spack install -f ...` form could handle relative paths:

```
cd /tmp
spack install -f bzip2.yaml
```

But with the change proposed in this PR, spack thinks that `bzip2` is some kind of namespace and expects `yaml` to refer to a package.  If we were to first treat the positional args as if they might be file paths, it would avoid this issue, but it's my understanding that doing things in that order would cause other problems.

**EDIT**

*This PR now simply documents that the above relative path must be prefixed with `./` in order to avoid the ambiguity.*

2. This PR doesn't support mixed mode specs, i.e. you can't pass as positional arguments some abstract spec strings along with some file paths, ala:

```
spack install pkgconf@1.4.2 /tmp/bzip2.yaml
```

There are currently some tests for mixed mode spec installation, see e.g. `test_install_mix_cli_and_files` in `lib/spack/spack/test/cmd/install.py`.  In that test, the mixed mode is achieved by passing positional arguments along with the `-f` option and supplying file paths.  Keeping the `-f` option where in the commands where it exists already could be a way to continue supporting mixed mode identification of specs, as well as provide backward compatibility for users expecting the `-f` option.

**EDIT**

*This PR now documents that the above mixed-mode approach to providing specs on the command line is not supported.*

3. What should happen if you have a `spec.yaml` from another system, and you accidentally try to install from it?  It may be seen as concrete due to being read from a yaml file, but the concretization may not be valid on the current system.  This may or may not already be an issue when using the `-f` option, however, so maybe isn't a big deal from the perspective of this PR.

scottwittenburg commented:
@tgamblin This is one of those three sprints we talked about last week.  I'm still working on it, and just made a PR so we'd have a place to talk about concrete/specific details with regard to the effort.

scottwittenburg commented:
@tgamblin As we discussed earlier today, I have created a new common command line argument `specs`, which simplified documenting the limitations on providing file paths to spec.yaml files as an alternative to clispecs for commands that operate on specs.  Let me know what you think if you get a chance.  Thanks!

tgamblin commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -143,15 +150,34 @@ def parse_specs(args, **kwargs):
-        raise SpackError(msg)
+        err_msg = msg
+
+    if err_msg:
```
Needs a comment that says this `if` is about reading files when spec parsing fails

tgamblin commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -143,15 +150,34 @@ def parse_specs(args, **kwargs):
+        try:
+            specs = []
+            for arg in args:
+                with open(arg, 'r') as fd:
```
Dont' really need the `'r'` as it's default

tgamblin commented on lib/spack/spack/cmd/__init__.py:
```diff
@@ -143,15 +150,34 @@ def parse_specs(args, **kwargs):
+            # exception at this point would indicate the spec *was* provided
+            # as a file path, and we should fail through so the user can see
+            # the underlying issue.
+            pass
```
I think you should probably raise here -- doesn't this mean that something was neither a valid spec or a filename?  If you skip it at this point I think you're allowing things like:

```
spack install <@#$> valid_filename.yaml
```

Am I missing something?

tgamblin commented on lib/spack/spack/test/cmd/install.py:
```diff
@@ -311,6 +311,26 @@ def test_install_from_file(spec, concretize, error_code, tmpdir):
+        spec.to_yaml(f)
+
+    # Test that we can install passing absolute path to spec yaml
+    install(str(specfile), fail_on_error=False)
```
In addition to testing the case where we provide filenames, there should be a case for mixed specs and filenames to ensure it fails -- you can mark it xfail, though probably better in this case to make sure it raises the exception you expect.

alalazo commented:
A high-level comment on the logic for parsing `specs`: if we don't allow mixed spec parsing maybe we could consider making the `-f` flag a global switch for a command? What I mean is something like:
```console
$ spack install -f bzip.yaml ... # Looks for files for all the specs

$ spack install bzip.yaml ... # Interpretes REMAINDER as a cli spec, so yaml package within namespace bzip
```
In this way:
1. We give the user control on whether they want to read from a file or concretize a cli spec, instead of relying on some automation. This removes the issues with the possibly unexpected logic pointed out in the description.
2. We'll still allow relative paths for the filenames, without the need to prepend them with `./`.

On the "practical" side this would be backward compatible with uses that install a single spec with `-f`:
```console
spack install -f bzip.yaml
```
On the implementation side this would also allow to keep logic and error-handling separate for the two cases, so it would solve ambiguites like https://github.com/spack/spack/pull/11103#discussion_r272811200 in the code. Would that make sense?

**Note**: currently we are allowing things like:
```console
$ spack install -f <file1> -f <file2> -f <file3> spec1 spec2 spec3
```
in a single command that won't be allowed anymore with this approach. Probably going the route this PR does it's not a big deal - I don't see many use cases for the line above, and in any case scripts that use it can be adapted - but I think it's worth pointing out that the proposed behavior is a restriction on what we can do currently.

tgamblin commented:
@alalazo: I like the suggestion to also standardize `-f`, but can that not be in a separate PR?

alalazo commented:
@tgamblin Sure, that's why I posted just a comment, not a request for change :smile:

healther commented:
@alalazo out of curiosity why do you do it in one CLI call? Would it make a difference to do it in multiple  `spack` calls (except for being more verbose)?

scottwittenburg commented:
I've rebased this topic, and I think I've addressed the issues raised in review here.  One thing I still don't like is the error handling logic in the mixed clispec/yaml case.  For example, issuing a `spack install pkgconf /tmp/bzip2.yaml` command results in:  

```
==> Error: No installed spec matches the hash: 'tmp'
```

There's a comment in that section of the code discussing the issue.   Maybe reviewers (@tgamblin, @scheibelp, @alalazo) have some thoughts on this.  Perhaps it would be cleaner to go with @alalazos idea to standarize the `-f` option in a general way across commands instead of this.

At any rate, the tests are passing, and if this seems like a reasonable change to make, I could work on getting the `codecov/patch` up a titch to make everything green.  Let me know.

scottwittenburg commented on lib/spack/spack/cmd/add.py:
```diff
@@ -17,8 +16,7 @@
 def setup_parser(subparser):
-    subparser.add_argument(
-        'specs', nargs=argparse.REMAINDER, help="specs of packages to add")
+    arguments.add_common_arguments(subparser, ['specs'])
```
@opadron Here is an example of how a command would make use of the common argument for specs so that either clispecs or yaml paths are accepted.

scottwittenburg commented:
We really need something like this in our pipelines, as in that case we always want to refer unambiguously to fully concrete specs by using `spec.yaml` files, and it seems cleaner to implement the functionality in a common way for any command that needs it.  This PR attempts to provide those benefits.  However, due to the ambiguities mentioned by @tgamblin and @alalazo, I'm not sure this PR is the right way to go.  Basically, it just seems too complicated to guess what the user intended in some of the error cases, and provide an appropriate error message.  

The other approach, mentioned above, would be to standardize the `-f` option across all the commands.  I think I may prefer that approach at this point, as it forces the user to be clear about how they want cli arguments treated, removing the ambiguity for us when handling those arguments.  

The only issue I can see with standardizing on `-f` is that there are currently a few commands that both 1) already take a `-f` argument to mean something else (most typically it's some kind of `force` option), and 2) need to take specs on the command line.  I've compiled a short list of those below:

```
activate             (force, activate without first activating dependencies)
buildcache create    (force, overwrite tarball if it exists)
buildcache install   (force, overwrite install directory if it exists)
buildcache list      (force, force new download of specs)
deactivate           (force, run deactivation even if spec is NOT currently activated)
mirror create        (file, file with specs of packages to put in mirror)
remove               (force, remove concretized spec (if any) immediately)
uninstall            (force, remove regardless of whether other packages or environments depend on this one)
```

Since `spack install ...` already takes a `-f` to mean `spec.yaml` files of specs to install, and since that's maybe spacks most commonly used command, it would seem to me that we should everywhere use `-f` to refer to `spec.yaml` files.  In that case, we would need to change the command listed above (maybe e.g. we don't provide a short option for `--force` anymore, but insist that users type it out in full).

scottwittenburg commented:
☝️ 

I don't want to spam all developers here, but I do want to draw some attention from key folks.  Feel free to ping other developers you think may have a stake.

@tgamblin @gartung @opadron @scheibelp @becker33 @adamjstewart

tgamblin commented:
Superseded by #12561.
